lib/cache/cache.go

```diff
@@ -548,38 +701,71 @@ func (c *Cache) fetchAndWatch(ctx context.Context, retry utils.Retry) error {
 			return trace.BadParameter("expected init event, got %v instead", event.Type)
 		}
 	}
-	err = c.fetch(ctx)
+	apply, err := c.fetch(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// apply will mutate cache, and possibly leave it in an invalid state
+	// if an error occurs, so ensure that cache is not read.
+	c.setReadOK(false)
+	err = apply()
```

Comment:
I understand from the implementation that `apply` implicitly uses the context specified to `fetch` but it looks  confusing w/o knowing that and given that it might fail with a context expired error could be unexpected - make it visually obvious by explicitly handing it the context?